Validation/HLTrigger/src/HLTGenResHistColl.cc

```diff
@@ -0,0 +1,303 @@
+  histConfigDefault0.addParameter<std::vector<double>>("vsBinLowEdges", defaultPtBinning);
+  std::vector<double> defaultResBinning;
+  defaultResBinning.reserve(151);
+  for (int i = 0; i < 151; i++) {
+    defaultResBinning.push_back(i * 0.01);
+  }
+  histConfigDefault0.addParameter<std::vector<double>>("resBinLowEdges", defaultResBinning);
+
+  histConfigDefaults.push_back(histConfigDefault0);
+  desc.addVPSet("histConfigs", histConfig, histConfigDefaults);
+
+  // defining single histConfig2D
+  edm::ParameterSetDescription histConfig2D;
+  histConfig2D.add<std::string>("vsVarX");
+  histConfig2D.add<std::string>("vsVarY");
+  histConfig2D.add<std::vector<double>>("binLowEdgesX");
+  histConfig2D.add<std::vector<double>>("binLowEdgesY");
+  histConfig2D.addVPSet(
+      "rangeCuts", VarRangeCut<HLTGenValObject>::makePSetDescription(), std::vector<edm::ParameterSet>());
+  // default set of histConfigs
+  std::vector<edm::ParameterSet> histConfigDefaults2D;
+  desc.addVPSet("histConfigs2D", histConfig2D, histConfigDefaults2D);
+
+  return desc;
+}
+
+bool HLTGenResHistColl::passFilterSelection(trigger::size_type objKey,
+                                            const trigger::TriggerEvent& triggerEvent) const {
+  if (andFilters_) {
+    for (const auto& filter : filters_) {
+      if (!passesFilter(objKey, triggerEvent, filter, hltProcessName_)) {
+        return false;
+      }
+    }
+    return true;
+  } else {
+    for (const auto& filter : filters_) {
+      if (passesFilter(objKey, triggerEvent, filter, hltProcessName_)) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
+
+// general hist booking function, receiving configurations for 1D and 2D hists and calling the respective functions
+void HLTGenResHistColl::bookHists(DQMStore::IBooker& iBooker) {
+  for (const auto& histConfig : histConfigs_)
+    book1D(iBooker, histConfig);
+  for (const auto& histConfig : histConfigs2D_)
+    book2D(iBooker, histConfig);
+}
+
+// histogram filling routine
+void HLTGenResHistColl::fillHists(const HLTGenValObject& obj, edm::Handle<trigger::TriggerEvent>& triggerEvent) {
+  // get the trigger objects of the collection
+  auto keyRange = getTrigObjIndicesOfCollection(*triggerEvent, collectionName_, hltProcessName_);
+  const trigger::TriggerObject* bestMatch = nullptr;
+  float bestDR2 = dR2limit_;
+  for (trigger::size_type key = keyRange.first; key < keyRange.second; ++key) {
+    if (passFilterSelection(key, *triggerEvent)) {
+      const trigger::TriggerObject objTrig = triggerEvent->getObjects().at(key);
+      if (isEventLevelVariable_) {
+        bestDR2 = 0;
+        bestMatch = &objTrig;
+        break;
+      } else {
+        float dR2 = reco::deltaR2(obj, objTrig);
+        if (dR2 < bestDR2) {
+          bestDR2 = dR2;
+          bestMatch = &objTrig;
+        }
+      }
+    }
+  }
+  // if(objType_=="ele") std::cout <<"objType "<<objType_<<" "<<obj.pt()<<" best "<<bestMatch<<std::endl;
```

Comment:
remove commented out line of code if not needed

Reply:
done